Use Go standard library slices package more extensively

tklauser opened the pull request:
Make use of some newly added functionality added in Go 1.23.

See commit messages for details.

tklauser committed:
Use Go standard library slices package instead of x/exp/slices

The former is available since Go 1.21 and the latter is merely a wrapper
around it when using Go ≥ 1.21. Use the standard library package
directly.

Signed-off-by: Tobias Klauser <tobias@cilium.io>

tklauser committed:
slices: change from sort functions to slices functions where feasible

Doing this because the slices functions are slightly faster and
easier to use. It also removes one dependency layer. This requires
changing SortedUniqueFunc to take the less function as a func(a, b T)
instead of func(i, j int) which is more idiomatic anyway.

Signed-off-by: Tobias Klauser <tobias@cilium.io>

gandro commented on pkg/ip/ip.go:
```diff
@@ -755,8 +755,8 @@ func PartitionCIDR(targetCIDR net.IPNet, excludeCIDR net.IPNet) ([]*net.IPNet, [
 		addrs,
-		func(i, j int) bool {
-			return addrs[i].Compare(addrs[j]) < 0
+		func(a, b netip.Addr) int {
+			return a.Compare(b)
```
How does this compile? There is no change in imports here

tklauser replied:
Note that `slices` here is not the Go stdlib package but `pkg/slices`, see 2nd commits which changes the signature of `SortedUniqueFunc`.

tklauser replied:
~That being said, we should probably import Go stdlib `slices` as such and alias `pkg/slices` instead.~ Scratch that, we don't import stdlib `slices` in that file.

gandro replied:
Ah, I was filtering files by my codeowners, so I didn't notice the signature of `SortedUniqueFunc` changing. Thanks for the explanation!

tklauser committed:
slices: zero out obsolete elements in Unique{,Func}

This allows GC to reclaim memory for these elements.

Signed-off-by: Tobias Klauser <tobias@cilium.io>

tklauser committed:
Use slices.Sort instead of sort.{Ints,Strings}

In recent Go versions sort.{Ints,Strings} is already implemented in
terms of slices.Sort. This change removes that indirection and a
dependency layer.

Signed-off-by: Tobias Klauser <tobias@cilium.io>

tklauser committed:
Use Contains* functions from Go standard library slices package

Use the generic Contains* functions from the slices Go standard library
package instead of open-coding them across the code base. This also
allows simplifying some code.

Signed-off-by: Tobias Klauser <tobias@cilium.io>

gandro reviewed (approved):
Looks good for my CODEOWNERs, thanks!

dylandreimerink reviewed (approved):
Files owned by teams I am on seem good!

sayboras reviewed (approved):
Looks good for envoy related files. Thanks.

giorio94 reviewed (approved):
/lgtm for my codeowners

harsimran-pabla reviewed (approved):
BGP changes looks good. Thank you 🙏

rolinh reviewed (approved):
Very nice cleanup, lgtm for Hubble!
